src/libsystemd/sd-daemon/sd-daemon.c

```diff
@@ -551,6 +552,77 @@ _public_ int sd_pid_notify_with_fds(
         return r;
 }
 
+_public_ int sd_notify_barrier(int timeout) {
+        union sockaddr_union sockaddr;
+        char byte = 0;
+        struct iovec iovec = {
+                .iov_base = &byte,
+                .iov_len = 1,
+        };
+        struct msghdr msghdr = {
+                .msg_iov = &iovec,
+                .msg_iovlen = 1,
+                .msg_name = &sockaddr,
+        };
+
+        _cleanup_close_pair_ int pipe_fd[2] = { -1, -1 };
+        _cleanup_close_ int fd = -1;
+        struct cmsghdr *cmsg = NULL;
+        const char *e;
+        int r;
+
+        e = getenv("NOTIFY_SOCKET");
+        if (!e)
+                return 0;
+
+        r = sockaddr_un_set_path(&sockaddr.un, e);
+        if (r < 0)
+                goto finish;
+        msghdr.msg_namelen = r;
+
+        fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0);
+        if (fd < 0) {
+                r = -errno;
+                goto finish;
+        }
+
+        if (pipe2(pipe_fd, 0) < 0) {
```

Comment:
should have at least O_CLOEXEC as flags.

(and probably also SOCK_NONBLOCK. Not because strictly necessary, but because it 'feels' more appropriate, given we never actually want to block on it other than with poll())